// hmrc-frontend/hmrc/components/user-research-banner/user-research-banner.js
(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory() :
  typeof define === 'function' && define.amd ? define('HMRCFrontend', factory) :
  (global.HMRCFrontend = factory());
}(this, (function () { 'use strict';

  // Based on https://github.com/alphagov/govuk_template_jinja
  var setCookie = function setCookie(name, value) {
    var options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
    var cookieString = "".concat(name, "=").concat(value, "; path=/");

    if (options.days) {
      var date = new Date();
      date.setTime(date.getTime() + options.days * 24 * 60 * 60 * 1000);
      cookieString = "".concat(cookieString, "; expires=").concat(date.toGMTString());
    }

    if (window.location.protocol === 'https:') {
      cookieString += '; Secure';
    }

    document.cookie = cookieString;
    return cookieString;
  };
  var getCookie = function getCookie(name) {
    var nameEQ = "".concat(name, "=");
    var cookies = document.cookie.split(';');

    for (var i = 0, len = cookies.length; i < len; i += 1) {
      var cookie = cookies[i];

      while (cookie.charAt(0) === ' ') {
        cookie = cookie.substring(1, cookie.length);
      }

      if (cookie.indexOf(nameEQ) === 0) {
        return decodeURIComponent(cookie.substring(nameEQ.length));
      }
    }

    return null;
  };

  function UserResearchBanner($module) {
    this.$module = $module;
    this.$closeLink = this.$module.querySelector('.hmrc-user-research-banner__close');
    this.cookieName = 'mdtpurr';
    this.cookieExpiryDays = 28;
  }

  UserResearchBanner.prototype.init = function init() {
    var cookieData = getCookie(this.cookieName);

    if (cookieData == null) {
      this.$module.classList.add('hmrc-user-research-banner--show');
      this.$closeLink.addEventListener('click', this.eventHandlers.noThanksClick.bind(this));
    }
  };

  UserResearchBanner.prototype.eventHandlers = {
    noThanksClick: function noThanksClick(event) {
      event.preventDefault();
      setCookie(this.cookieName, 'suppress_for_all_services', {
        days: this.cookieExpiryDays
      });
      this.$module.classList.remove('hmrc-user-research-banner--show');
    }
  };

  return UserResearchBanner;

})));
