// hmrc-frontend/hmrc/vendor/polyfills/Element/prototype/dataset.js
(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? factory() :
  typeof define === 'function' && define.amd ? define('HMRCFrontend', factory) :
  (factory());
}(this, (function () { 'use strict';

  function _typeof(obj) {
    "@babel/helpers - typeof";

    if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {
      _typeof = function (obj) {
        return typeof obj;
      };
    } else {
      _typeof = function (obj) {
        return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj;
      };
    }

    return _typeof(obj);
  }

  var commonjsGlobal = typeof globalThis !== 'undefined' ? globalThis : typeof window !== 'undefined' ? window : typeof global !== 'undefined' ? global : typeof self !== 'undefined' ? self : {};

  function createCommonjsModule(fn, module) {
  	return module = { exports: {} }, fn(module, module.exports), module.exports;
  }

  var defineProperty = createCommonjsModule(function (module, exports) {
  (function (global, factory) {
    factory();
  })(commonjsGlobal, function () {

    (function (undefined) {
      // Detection from https://github.com/Financial-Times/polyfill-service/blob/master/packages/polyfill-library/polyfills/Object/defineProperty/detect.js
      var detect = // In IE8, defineProperty could only act on DOM elements, so full support
      // for the feature requires the ability to set a property on an arbitrary object
      'defineProperty' in Object && function () {
        try {
          var a = {};
          Object.defineProperty(a, 'test', {
            value: 42
          });
          return true;
        } catch (e) {
          return false;
        }
      }();

      if (detect) return; // Polyfill from https://cdn.polyfill.io/v2/polyfill.js?features=Object.defineProperty&flags=always

      (function (nativeDefineProperty) {
        var supportsAccessors = Object.prototype.hasOwnProperty('__defineGetter__');
        var ERR_ACCESSORS_NOT_SUPPORTED = 'Getters & setters cannot be defined on this javascript engine';
        var ERR_VALUE_ACCESSORS = 'A property cannot both have accessors and be writable or have a value';

        Object.defineProperty = function defineProperty(object, property, descriptor) {
          // Where native support exists, assume it
          if (nativeDefineProperty && (object === window || object === document || object === Element.prototype || object instanceof Element)) {
            return nativeDefineProperty(object, property, descriptor);
          }

          if (object === null || !(object instanceof Object || typeof object === 'object')) {
            throw new TypeError('Object.defineProperty called on non-object');
          }

          if (!(descriptor instanceof Object)) {
            throw new TypeError('Property description must be an object');
          }

          var propertyString = String(property);
          var hasValueOrWritable = 'value' in descriptor || 'writable' in descriptor;
          var getterType = 'get' in descriptor && typeof descriptor.get;
          var setterType = 'set' in descriptor && typeof descriptor.set; // handle descriptor.get

          if (getterType) {
            if (getterType !== 'function') {
              throw new TypeError('Getter must be a function');
            }

            if (!supportsAccessors) {
              throw new TypeError(ERR_ACCESSORS_NOT_SUPPORTED);
            }

            if (hasValueOrWritable) {
              throw new TypeError(ERR_VALUE_ACCESSORS);
            }

            Object.__defineGetter__.call(object, propertyString, descriptor.get);
          } else {
            object[propertyString] = descriptor.value;
          } // handle descriptor.set


          if (setterType) {
            if (setterType !== 'function') {
              throw new TypeError('Setter must be a function');
            }

            if (!supportsAccessors) {
              throw new TypeError(ERR_ACCESSORS_NOT_SUPPORTED);
            }

            if (hasValueOrWritable) {
              throw new TypeError(ERR_VALUE_ACCESSORS);
            }

            Object.__defineSetter__.call(object, propertyString, descriptor.set);
          } // OK to define value unconditionally - if a getter has been specified as well, an error would be thrown above


          if ('value' in descriptor) {
            object[propertyString] = descriptor.value;
          }

          return object;
        };
      })(Object.defineProperty);
    }).call('object' === typeof window && window || 'object' === typeof self && self || 'object' === typeof commonjsGlobal && commonjsGlobal || {});
  });
  });

  var bind = createCommonjsModule(function (module, exports) {
  (function (global, factory) {
    factory();
  })(commonjsGlobal, function () {

    (function (undefined) {
      // Detection from https://github.com/Financial-Times/polyfill-service/blob/master/packages/polyfill-library/polyfills/Object/defineProperty/detect.js
      var detect = // In IE8, defineProperty could only act on DOM elements, so full support
      // for the feature requires the ability to set a property on an arbitrary object
      'defineProperty' in Object && function () {
        try {
          var a = {};
          Object.defineProperty(a, 'test', {
            value: 42
          });
          return true;
        } catch (e) {
          return false;
        }
      }();

      if (detect) return; // Polyfill from https://cdn.polyfill.io/v2/polyfill.js?features=Object.defineProperty&flags=always

      (function (nativeDefineProperty) {
        var supportsAccessors = Object.prototype.hasOwnProperty('__defineGetter__');
        var ERR_ACCESSORS_NOT_SUPPORTED = 'Getters & setters cannot be defined on this javascript engine';
        var ERR_VALUE_ACCESSORS = 'A property cannot both have accessors and be writable or have a value';

        Object.defineProperty = function defineProperty(object, property, descriptor) {
          // Where native support exists, assume it
          if (nativeDefineProperty && (object === window || object === document || object === Element.prototype || object instanceof Element)) {
            return nativeDefineProperty(object, property, descriptor);
          }

          if (object === null || !(object instanceof Object || typeof object === 'object')) {
            throw new TypeError('Object.defineProperty called on non-object');
          }

          if (!(descriptor instanceof Object)) {
            throw new TypeError('Property description must be an object');
          }

          var propertyString = String(property);
          var hasValueOrWritable = 'value' in descriptor || 'writable' in descriptor;
          var getterType = 'get' in descriptor && typeof descriptor.get;
          var setterType = 'set' in descriptor && typeof descriptor.set; // handle descriptor.get

          if (getterType) {
            if (getterType !== 'function') {
              throw new TypeError('Getter must be a function');
            }

            if (!supportsAccessors) {
              throw new TypeError(ERR_ACCESSORS_NOT_SUPPORTED);
            }

            if (hasValueOrWritable) {
              throw new TypeError(ERR_VALUE_ACCESSORS);
            }

            Object.__defineGetter__.call(object, propertyString, descriptor.get);
          } else {
            object[propertyString] = descriptor.value;
          } // handle descriptor.set


          if (setterType) {
            if (setterType !== 'function') {
              throw new TypeError('Setter must be a function');
            }

            if (!supportsAccessors) {
              throw new TypeError(ERR_ACCESSORS_NOT_SUPPORTED);
            }

            if (hasValueOrWritable) {
              throw new TypeError(ERR_VALUE_ACCESSORS);
            }

            Object.__defineSetter__.call(object, propertyString, descriptor.set);
          } // OK to define value unconditionally - if a getter has been specified as well, an error would be thrown above


          if ('value' in descriptor) {
            object[propertyString] = descriptor.value;
          }

          return object;
        };
      })(Object.defineProperty);
    }).call('object' === typeof window && window || 'object' === typeof self && self || 'object' === typeof commonjsGlobal && commonjsGlobal || {});
    (function (undefined) {
      // Detection from https://github.com/Financial-Times/polyfill-service/blob/master/packages/polyfill-library/polyfills/Function/prototype/bind/detect.js
      var detect = ('bind' in Function.prototype);
      if (detect) return; // Polyfill from https://cdn.polyfill.io/v2/polyfill.js?features=Function.prototype.bind&flags=always

      Object.defineProperty(Function.prototype, 'bind', {
        value: function bind(that) {
          // .length is 1
          // add necessary es5-shim utilities
          var $Array = Array;
          var $Object = Object;
          var ObjectPrototype = $Object.prototype;
          var ArrayPrototype = $Array.prototype;

          var Empty = function Empty() {};

          var to_string = ObjectPrototype.toString;
          var hasToStringTag = typeof Symbol === 'function' && typeof Symbol.toStringTag === 'symbol';
          var isCallable;
          /* inlined from https://npmjs.com/is-callable */

          var fnToStr = Function.prototype.toString,
              tryFunctionObject = function tryFunctionObject(value) {
            try {
              fnToStr.call(value);
              return true;
            } catch (e) {
              return false;
            }
          },
              fnClass = '[object Function]',
              genClass = '[object GeneratorFunction]';

          isCallable = function isCallable(value) {
            if (typeof value !== 'function') {
              return false;
            }

            if (hasToStringTag) {
              return tryFunctionObject(value);
            }

            var strClass = to_string.call(value);
            return strClass === fnClass || strClass === genClass;
          };

          var array_slice = ArrayPrototype.slice;
          var array_concat = ArrayPrototype.concat;
          var array_push = ArrayPrototype.push;
          var max = Math.max; // /add necessary es5-shim utilities
          // 1. Let Target be the this value.

          var target = this; // 2. If IsCallable(Target) is false, throw a TypeError exception.

          if (!isCallable(target)) {
            throw new TypeError('Function.prototype.bind called on incompatible ' + target);
          } // 3. Let A be a new (possibly empty) internal list of all of the
          //   argument values provided after thisArg (arg1, arg2 etc), in order.
          // XXX slicedArgs will stand in for "A" if used


          var args = array_slice.call(arguments, 1); // for normal call
          // 4. Let F be a new native ECMAScript object.
          // 11. Set the [[Prototype]] internal property of F to the standard
          //   built-in Function prototype object as specified in 15.3.3.1.
          // 12. Set the [[Call]] internal property of F as described in
          //   15.3.4.5.1.
          // 13. Set the [[Construct]] internal property of F as described in
          //   15.3.4.5.2.
          // 14. Set the [[HasInstance]] internal property of F as described in
          //   15.3.4.5.3.

          var bound;

          var binder = function () {
            if (this instanceof bound) {
              // 15.3.4.5.2 [[Construct]]
              // When the [[Construct]] internal method of a function object,
              // F that was created using the bind function is called with a
              // list of arguments ExtraArgs, the following steps are taken:
              // 1. Let target be the value of F's [[TargetFunction]]
              //   internal property.
              // 2. If target has no [[Construct]] internal method, a
              //   TypeError exception is thrown.
              // 3. Let boundArgs be the value of F's [[BoundArgs]] internal
              //   property.
              // 4. Let args be a new list containing the same values as the
              //   list boundArgs in the same order followed by the same
              //   values as the list ExtraArgs in the same order.
              // 5. Return the result of calling the [[Construct]] internal
              //   method of target providing args as the arguments.
              var result = target.apply(this, array_concat.call(args, array_slice.call(arguments)));

              if ($Object(result) === result) {
                return result;
              }

              return this;
            } else {
              // 15.3.4.5.1 [[Call]]
              // When the [[Call]] internal method of a function object, F,
              // which was created using the bind function is called with a
              // this value and a list of arguments ExtraArgs, the following
              // steps are taken:
              // 1. Let boundArgs be the value of F's [[BoundArgs]] internal
              //   property.
              // 2. Let boundThis be the value of F's [[BoundThis]] internal
              //   property.
              // 3. Let target be the value of F's [[TargetFunction]] internal
              //   property.
              // 4. Let args be a new list containing the same values as the
              //   list boundArgs in the same order followed by the same
              //   values as the list ExtraArgs in the same order.
              // 5. Return the result of calling the [[Call]] internal method
              //   of target providing boundThis as the this value and
              //   providing args as the arguments.
              // equiv: target.call(this, ...boundArgs, ...args)
              return target.apply(that, array_concat.call(args, array_slice.call(arguments)));
            }
          }; // 15. If the [[Class]] internal property of Target is "Function", then
          //     a. Let L be the length property of Target minus the length of A.
          //     b. Set the length own property of F to either 0 or L, whichever is
          //       larger.
          // 16. Else set the length own property of F to 0.


          var boundLength = max(0, target.length - args.length); // 17. Set the attributes of the length own property of F to the values
          //   specified in 15.3.5.1.

          var boundArgs = [];

          for (var i = 0; i < boundLength; i++) {
            array_push.call(boundArgs, '$' + i);
          } // XXX Build a dynamic function with desired amount of arguments is the only
          // way to set the length property of a function.
          // In environments where Content Security Policies enabled (Chrome extensions,
          // for ex.) all use of eval or Function costructor throws an exception.
          // However in all of these environments Function.prototype.bind exists
          // and so this code will never be executed.


          bound = Function('binder', 'return function (' + boundArgs.join(',') + '){ return binder.apply(this, arguments); }')(binder);

          if (target.prototype) {
            Empty.prototype = target.prototype;
            bound.prototype = new Empty(); // Clean up dangling references.

            Empty.prototype = null;
          } // TODO
          // 18. Set the [[Extensible]] internal property of F to true.
          // TODO
          // 19. Let thrower be the [[ThrowTypeError]] function Object (13.2.3).
          // 20. Call the [[DefineOwnProperty]] internal method of F with
          //   arguments "caller", PropertyDescriptor {[[Get]]: thrower, [[Set]]:
          //   thrower, [[Enumerable]]: false, [[Configurable]]: false}, and
          //   false.
          // 21. Call the [[DefineOwnProperty]] internal method of F with
          //   arguments "arguments", PropertyDescriptor {[[Get]]: thrower,
          //   [[Set]]: thrower, [[Enumerable]]: false, [[Configurable]]: false},
          //   and false.
          // TODO
          // NOTE Function objects created using Function.prototype.bind do not
          // have a prototype property or the [[Code]], [[FormalParameters]], and
          // [[Scope]] internal properties.
          // XXX can't delete prototype in pure-js.
          // 22. Return F.


          return bound;
        }
      });
    }).call('object' === typeof window && window || 'object' === typeof self && self || 'object' === typeof commonjsGlobal && commonjsGlobal || {});
  });
  });

  (function (undefined) {
    // Detection from https://github.com/Financial-Times/polyfill-library/blob/master/polyfills/Object/getOwnPropertyDescriptor/detect.js
    var detect = 'getOwnPropertyDescriptor' in Object && typeof Object.getOwnPropertyDescriptor === 'function' && function () {
      try {
        var object = {};
        object.test = 0;
        return Object.getOwnPropertyDescriptor(object, "test").value === 0;
      } catch (exception) {
        return false;
      }
    }();

    if (detect) return; // Polyfill from https://cdn.polyfill.io/v2/polyfill.js?features=Object.getOwnPropertyDescriptor&flags=always

    (function () {
      var call = Function.prototype.call;
      var prototypeOfObject = Object.prototype;
      var owns = call.bind(prototypeOfObject.hasOwnProperty);
      var lookupGetter;
      var lookupSetter;
      var supportsAccessors;

      if (supportsAccessors = owns(prototypeOfObject, "__defineGetter__")) {
        lookupGetter = call.bind(prototypeOfObject.__lookupGetter__);
        lookupSetter = call.bind(prototypeOfObject.__lookupSetter__);
      }

      function doesGetOwnPropertyDescriptorWork(object) {
        try {
          object.sentinel = 0;
          return Object.getOwnPropertyDescriptor(object, "sentinel").value === 0;
        } catch (exception) {// returns falsy
        }
      } // check whether getOwnPropertyDescriptor works if it's given. Otherwise,
      // shim partially.


      if (Object.defineProperty) {
        var getOwnPropertyDescriptorWorksOnObject = doesGetOwnPropertyDescriptorWork({});
        var getOwnPropertyDescriptorWorksOnDom = typeof document == "undefined" || doesGetOwnPropertyDescriptorWork(document.createElement("div"));

        if (!getOwnPropertyDescriptorWorksOnDom || !getOwnPropertyDescriptorWorksOnObject) {
          var getOwnPropertyDescriptorFallback = Object.getOwnPropertyDescriptor;
        }
      }

      if (!Object.getOwnPropertyDescriptor || getOwnPropertyDescriptorFallback) {
        var ERR_NON_OBJECT = "Object.getOwnPropertyDescriptor called on a non-object: ";

        Object.getOwnPropertyDescriptor = function getOwnPropertyDescriptor(object, property) {
          if (_typeof(object) != "object" && typeof object != "function" || object === null) {
            throw new TypeError(ERR_NON_OBJECT + object);
          } // make a valiant attempt to use the real getOwnPropertyDescriptor
          // for I8's DOM elements.


          if (getOwnPropertyDescriptorFallback) {
            try {
              return getOwnPropertyDescriptorFallback.call(Object, object, property);
            } catch (exception) {// try the shim if the real one doesn't work
            }
          } // If object does not owns property return undefined immediately.


          if (!owns(object, property)) {
            return;
          } // If object has a property then it's for sure both `enumerable` and
          // `configurable`.


          var descriptor = {
            enumerable: true,
            configurable: true
          }; // If JS engine supports accessor properties then property may be a
          // getter or setter.

          if (supportsAccessors) {
            // Unfortunately `__lookupGetter__` will return a getter even
            // if object has own non getter property along with a same named
            // inherited getter. To avoid misbehavior we temporary remove
            // `__proto__` so that `__lookupGetter__` will return getter only
            // if it's owned by an object.
            var prototype = object.__proto__;
            object.__proto__ = prototypeOfObject;
            var getter = lookupGetter(object, property);
            var setter = lookupSetter(object, property); // Once we have getter and setter we can put values back.

            object.__proto__ = prototype;

            if (getter || setter) {
              if (getter) {
                descriptor.get = getter;
              }

              if (setter) {
                descriptor.set = setter;
              } // If it was accessor property we're done and return here
              // in order to avoid adding `value` to the descriptor.


              return descriptor;
            }
          } // If we got this far we know that object has an own property that is
          // not an accessor so we set it as a value and return descriptor.


          descriptor.value = object[property];
          descriptor.writable = true;
          return descriptor;
        };
      }
    })();
  }).call('object' === (typeof window === "undefined" ? "undefined" : _typeof(window)) && window || 'object' === (typeof self === "undefined" ? "undefined" : _typeof(self)) && self || 'object' === (typeof global === "undefined" ? "undefined" : _typeof(global)) && global || {});

  var Document$1 = createCommonjsModule(function (module, exports) {
  (function (global, factory) {
    factory();
  })(commonjsGlobal, function () {

    (function (undefined) {
      // Detection from https://github.com/Financial-Times/polyfill-service/blob/master/packages/polyfill-library/polyfills/Document/detect.js
      var detect = ("Document" in this);
      if (detect) return; // Polyfill from https://cdn.polyfill.io/v2/polyfill.js?features=Document&flags=always

      if (typeof WorkerGlobalScope === "undefined" && typeof importScripts !== "function") {
        if (this.HTMLDocument) {
          // IE8
          // HTMLDocument is an extension of Document.  If the browser has HTMLDocument but not Document, the former will suffice as an alias for the latter.
          this.Document = this.HTMLDocument;
        } else {
          // Create an empty function to act as the missing constructor for the document object, attach the document object as its prototype.  The function needs to be anonymous else it is hoisted and causes the feature detect to prematurely pass, preventing the assignments below being made.
          this.Document = this.HTMLDocument = document.constructor = new Function('return function Document() {}')();
          this.Document.prototype = document;
        }
      }
    }).call('object' === typeof window && window || 'object' === typeof self && self || 'object' === typeof commonjsGlobal && commonjsGlobal || {});
  });
  });

  var Element_1 = createCommonjsModule(function (module, exports) {
  (function (global, factory) {
    factory();
  })(commonjsGlobal, function () {

    (function (undefined) {
      // Detection from https://github.com/Financial-Times/polyfill-service/blob/master/packages/polyfill-library/polyfills/Document/detect.js
      var detect = ("Document" in this);
      if (detect) return; // Polyfill from https://cdn.polyfill.io/v2/polyfill.js?features=Document&flags=always

      if (typeof WorkerGlobalScope === "undefined" && typeof importScripts !== "function") {
        if (this.HTMLDocument) {
          // IE8
          // HTMLDocument is an extension of Document.  If the browser has HTMLDocument but not Document, the former will suffice as an alias for the latter.
          this.Document = this.HTMLDocument;
        } else {
          // Create an empty function to act as the missing constructor for the document object, attach the document object as its prototype.  The function needs to be anonymous else it is hoisted and causes the feature detect to prematurely pass, preventing the assignments below being made.
          this.Document = this.HTMLDocument = document.constructor = new Function('return function Document() {}')();
          this.Document.prototype = document;
        }
      }
    }).call('object' === typeof window && window || 'object' === typeof self && self || 'object' === typeof commonjsGlobal && commonjsGlobal || {});
    (function (undefined) {
      // Detection from https://github.com/Financial-Times/polyfill-service/blob/master/packages/polyfill-library/polyfills/Element/detect.js
      var detect = 'Element' in this && 'HTMLElement' in this;
      if (detect) return; // Polyfill from https://cdn.polyfill.io/v2/polyfill.js?features=Element&flags=always

      (function () {
        // IE8
        if (window.Element && !window.HTMLElement) {
          window.HTMLElement = window.Element;
          return;
        } // create Element constructor


        window.Element = window.HTMLElement = new Function('return function Element() {}')(); // generate sandboxed iframe

        var vbody = document.appendChild(document.createElement('body'));
        var frame = vbody.appendChild(document.createElement('iframe')); // use sandboxed iframe to replicate Element functionality

        var frameDocument = frame.contentWindow.document;
        var prototype = Element.prototype = frameDocument.appendChild(frameDocument.createElement('*'));
        var cache = {}; // polyfill Element.prototype on an element

        var shiv = function (element, deep) {
          var childNodes = element.childNodes || [],
              index = -1,
              key,
              value,
              childNode;

          if (element.nodeType === 1 && element.constructor !== Element) {
            element.constructor = Element;

            for (key in cache) {
              value = cache[key];
              element[key] = value;
            }
          }

          while (childNode = deep && childNodes[++index]) {
            shiv(childNode, deep);
          }

          return element;
        };

        var elements = document.getElementsByTagName('*');
        var nativeCreateElement = document.createElement;
        var interval;
        var loopLimit = 100;
        prototype.attachEvent('onpropertychange', function (event) {
          var propertyName = event.propertyName,
              nonValue = !cache.hasOwnProperty(propertyName),
              newValue = prototype[propertyName],
              oldValue = cache[propertyName],
              index = -1,
              element;

          while (element = elements[++index]) {
            if (element.nodeType === 1) {
              if (nonValue || element[propertyName] === oldValue) {
                element[propertyName] = newValue;
              }
            }
          }

          cache[propertyName] = newValue;
        });
        prototype.constructor = Element;

        if (!prototype.hasAttribute) {
          // <Element>.hasAttribute
          prototype.hasAttribute = function hasAttribute(name) {
            return this.getAttribute(name) !== null;
          };
        } // Apply Element prototype to the pre-existing DOM as soon as the body element appears.


        function bodyCheck() {
          if (!loopLimit--) clearTimeout(interval);

          if (document.body && !document.body.prototype && /(complete|interactive)/.test(document.readyState)) {
            shiv(document, true);
            if (interval && document.body.prototype) clearTimeout(interval);
            return !!document.body.prototype;
          }

          return false;
        }

        if (!bodyCheck()) {
          document.onreadystatechange = bodyCheck;
          interval = setInterval(bodyCheck, 25);
        } // Apply to any new elements created after load


        document.createElement = function createElement(nodeName) {
          var element = nativeCreateElement(String(nodeName).toLowerCase());
          return shiv(element);
        }; // remove sandboxed iframe


        document.removeChild(vbody);
      })();
    }).call('object' === typeof window && window || 'object' === typeof self && self || 'object' === typeof commonjsGlobal && commonjsGlobal || {});
  });
  });

  (function (undefined) {
    // Detection from https://github.com/Financial-Times/polyfill-library/blob/master/polyfills/document/querySelector/detect.js
    var detect = 'document' in this && 'querySelector' in this.document;
    if (detect) return; // Polyfill from https://cdn.polyfill.io/v2/polyfill.js?features=Event&flags=always

    (function () {
      var head = document.getElementsByTagName('head')[0];

      function getElementsByQuery(node, selector, one) {
        var generator = document.createElement('div'),
            id = 'qsa' + String(Math.random()).slice(3),
            style,
            elements;
        generator.innerHTML = 'x<style>' + selector + '{qsa:' + id + ';}';
        style = head.appendChild(generator.lastChild);
        elements = getElements(node, selector, one, id);
        head.removeChild(style);
        return one ? elements[0] : elements;
      }

      function getElements(node, selector, one, id) {
        var validNode = /1|9/.test(node.nodeType),
            childNodes = node.childNodes,
            elements = [],
            index = -1,
            childNode;

        if (validNode && node.currentStyle && node.currentStyle.qsa === id) {
          if (elements.push(node) && one) {
            return elements;
          }
        }

        while (childNode = childNodes[++index]) {
          elements = elements.concat(getElements(childNode, selector, one, id));

          if (one && elements.length) {
            return elements;
          }
        }

        return elements;
      }

      Document.prototype.querySelector = Element.prototype.querySelector = function querySelectorAll(selector) {
        return getElementsByQuery(this, selector, true);
      };

      Document.prototype.querySelectorAll = Element.prototype.querySelectorAll = function querySelectorAll(selector) {
        return getElementsByQuery(this, selector, false);
      };
    })();
  }).call('object' === (typeof window === "undefined" ? "undefined" : _typeof(window)) && window || 'object' === (typeof self === "undefined" ? "undefined" : _typeof(self)) && self || 'object' === (typeof global === "undefined" ? "undefined" : _typeof(global)) && global || {});

  (function (undefined) {
    // Detection from https://github.com/Financial-Times/polyfill-library/blob/master/polyfills/Element/prototype/dataset/detect.js
    (function () {
      if (!document.documentElement.dataset) {
        return false;
      }

      var el = document.createElement('div');
      el.setAttribute("data-a-b", "c");
      return el.dataset && el.dataset.aB == "c";
    })(); // Polyfill from https://cdn.polyfill.io/v2/polyfill.js?features=Element.prototype.dataset&flags=always


    Object.defineProperty(Element.prototype, 'dataset', {
      get: function get() {
        var element = this;
        var attributes = this.attributes;
        var map = {};

        for (var i = 0; i < attributes.length; i++) {
          var attribute = attributes[i];

          if (attribute && attribute.name && /^data-\w[\w\-]*$/.test(attribute.name)) {
            var name = attribute.name;
            var value = attribute.value;
            var propName = name.substr(5).replace(/-./g, function (prop) {
              return prop.charAt(1).toUpperCase();
            });
            Object.defineProperty(map, propName, {
              enumerable: this.enumerable,
              get: function () {
                return this.value;
              }.bind({
                value: value || ''
              }),
              set: function setter(name, value) {
                if (typeof value !== 'undefined') {
                  this.setAttribute(name, value);
                } else {
                  this.removeAttribute(name);
                }
              }.bind(element, name)
            });
          }
        }

        return map;
      }
    });
  }).call('object' === (typeof window === "undefined" ? "undefined" : _typeof(window)) && window || 'object' === (typeof self === "undefined" ? "undefined" : _typeof(self)) && self || 'object' === (typeof global === "undefined" ? "undefined" : _typeof(global)) && global || {});

})));
